### ISSUE TITLE
Incorrect supply/minerals for multiple frames when training

### ISSUE THREAD

**Issue by Lamarth** [bwapi/bwapi#652]
Write a simple 4-pool zergling bot in C++, and observed with:

> myfile << Broodwar->getFrameCount() << "," << Broodwar->self()->supplyUsed() << "," << Broodwar->self()->minerals() << "," << Broodwar->self()->gas() << std::endl;

The latency frames is 3 with my LAN between VMs. Every latency frames this simple bot will act and there will be a second log line after the action. There are quite a lot of things wrong with the following:

> 2345,6,202,0
> 2346,6,202,0
> 2346,9,2,0
> 2347,9,2,0
> 2348,9,2,0
> 2349,9,-148,0
> 2349,9,-148,0
> 2350,15,-148,0
> 2351,15,-148,0
> 2352,15,52,0
> 2352,15,52,0
> 2353,12,52,0
> 2354,12,52,0
1. 200 minerals subtracted - for 3 pairs of lings? I suspect 4 calls were made, but it was definitely one for each larva in Broodwar->self()->getUnits().
2. 3 supply is used for 3 sets of lings, instead of 6
3. 3 frames later, the actual change in minerals arrives from the server, but the original negative isn't removed
4. the next frame, the supply change arrives from the server, again with the original supply offset still in effect

Problem number 1 doesn't seem to happen every time.
Problem number 2 appears to be related to using train or build rather than morph.
Problems number 3 and 4 each linger for 3 frames, and are the ones that really mess up algorithms (e.g. for building supply when running low).

Code for this simple bot is available if desired. There are similar but different problems with my F# bot that I have not thoroughly examined at this stage.

[recordMorph.txt](https://github.com/bwapi/bwapi/files/196476/recordMorph.txt)
[recordTrain.txt](https://github.com/bwapi/bwapi/files/196477/recordTrain.txt)

**Comment by heinermann** [comment]
Seems like an issue with latency compensation.

**Comment by Lamarth** [comment]
I agree that latency compensation set to 6 frames while latency is set to 3 would cause the bulk of the problem.

**Comment by heinermann** [comment]
Which version of BWAPI was used?

**Comment by Lamarth** [comment]
4.1.2

**Comment by jaj22** [comment]
Spotted a likely bug here. `CommandTemp<>::execute()` (which appears to contain the bulk of the latency compensation code) uses `getLatency()` throughout rather than `getLatencyFrames()`. Pretty sure this isn't intentional.

[https://github.com/bwapi/bwapi/blob/master/bwapi/BWAPI/Source/BWAPI/CommandTemp.h](https://github.com/bwapi/bwapi/blob/master/bwapi/BWAPI/Source/BWAPI/CommandTemp.h)

This is probably the cause of a difference in supplyUsed behaviour between SSCAIT vs local UDP. Both respond with getLatencyFrames() == 3, but SSCAIT may have a different value for getLatency().

**Comment by Cmccrave** [comment]
+1 to what jaj22 said, my early build order is based off supply and consistently 1 supply early on SSCAIT stream, but is fine locally.

**Comment by heinermann** [comment]
Thanks for diving into it. Will try to verify if changing it to getLatencyFrames will fix it.

**Comment by dshnightmare** [comment]
I find this problem too in version 4.1.2

**Comment by N00byEdge** [comment]
Issue still present in 4.2.0 and I assume it's still not fixed?

**Comment by xatian** [comment]
This problem is actually easy to reproduce:
```
printf ("Frame[%04d] mineralCount: %d\n", BWAPI::Broodwar->getFrameCount (), BWAPI::Broodwar->self ()->minerals ());
if (BWAPI::Broodwar->getFrameCount () == 0) {
    for (BWAPI::Unit unit : BWAPI::Broodwar->self ()->getUnits ())
        if (unit->getType () == BWAPI::UnitTypes::Zerg_Larva) {
            unit->morph (BWAPI::UnitTypes::Zerg_Drone);
            return;
        }
}
```
Output (Latency-compensation **enabled** - Version 4.2 calling morph/train or build):
```
Frame[0000] mineralCount: 50
Frame[0001] mineralCount: 0
Frame[0002] mineralCount: 0
Frame[0003] mineralCount: -50
Frame[0004] mineralCount: -50
Frame[0005] mineralCount: 0
Frame[0006] mineralCount: 0
Frame[0007] mineralCount: 0
```
Output (Latency-compensation **disabled** - Version 4.2 calling morph/train or build):
```
Frame[0000] mineralCount: 50
Frame[0001] mineralCount: 50
Frame[0002] mineralCount: 50
Frame[0003] mineralCount: 0
Frame[0004] mineralCount: 0
Frame[0005] mineralCount: 0
Frame[0006] mineralCount: 0
Frame[0007] mineralCount: 0
```
This only happens on LAN games where getLatencyFrames>2.

In my opinion the latency-compensation-code cannot be the problem since the price of the drone is deducted correctly. Something goes wrong on frame 3 and 5.


Another (maybe) helpful information:
With Version 4.1 calling train worked fine (no negative mineral-count) but morph and build was even worse:
```
Frame[0000] mineralCount: 50
Frame[0001] mineralCount: 0
Frame[0002] mineralCount: 0
Frame[0003] mineralCount: 0
Frame[0004] mineralCount: 50
Frame[0005] mineralCount: -50
Frame[0006] mineralCount: -50
Frame[0007] mineralCount: 0
```

**Comment by N00byEdge** [comment]
@xatian This is a latency compensation issue. The timing for latency compensation is thrown off by turn sizes > 2. I am somewhat slowly rewriting latcom to "fix" this.

**Comment by xatian** [comment]
@N00byEdge I don't understand.
I tried to trace the latcom code --> CommandTemp::execute is called just once which reduces the mineral-count down to 0. I wasn't able to figure out where in the code the mineral-count is set to -50 and back to 0 again but I am certain it is not in CommandTemp::execute.
Is there more latcom-code which is not part of CommandTemp?

**Comment by N00byEdge** [comment]
@xatian Yes. The `CommandTemp` is placed into a command buffer. On each frame, the most recent game data is pulled and all commands in the buffer are executed again. The timing within this buffer is not working correctly.

**Comment by xatian** [comment]
Can you please tell me where in the code this is happening ...
I added a printf at the start of CommandTemp::execute () and it triggered just once.

**Comment by jaj22** [comment]
CommandTemp::execute(n) is called once immediately when the command is buffered, and then once per buffered command per frame. Unless you compile your own BWAPI DLL you'll only see the immediate call. See GameImpl::applyLatencyCompensation.

If you want mineral and supply counts to be usable, disable latency compensation with `Game::setLatCom(false)`. You can avoid double-ordering units by comparing Unit::getLastCommandFrame to Game::getLatencyFrames, although it needs some care with turn size >1.

**Comment by xatian** [comment]
@jaj22 Thanks for your comment ... took some time but I understand now :-)
Curious design to do LatCom in the Server ... it's stuff to help the client (the bot) so imho it should be implemented, calculated and activated in the client who needs it.

But that makes me wonder:
If a tournament is hosted via "local PC" (so there is just one BWAPI-Server) -->
Does calling `Game::setLatCom(false)` affect the other bot too? If that is the case (and it looks that way since it's a game setting and not a player setting) one cannot simply change this setting --> it would break the other bot.
And Second ... that would mean that LatCom is applied to the  enemy units too and my bot can "see" what enemy units are doing before the latency is over?

Furthermore I looked into your suggestion and it is of course doable but I don't like it because:
It is more work **and** it's a disadvantage in some situations where the client won't accept commands due to latency. For example when cancelling a morph without LatCom you need to wait to get the resources refunded before you can build something else. With LatCom you can build immediately saving 1xLatencyFrames time. Yeah ... it's not much but more work for a disadvantage --> which is bad.

@N00byEdge You said you are going to fix this? :-)
_(I would of course volunteer as well but I don't know if it is even worth to fix with 5.0 coming up.)_

**Comment by N00byEdge** [comment]
> But that makes me wonder:
> If a tournament is hosted via "local PC" (so there is just one BWAPI-Server) -->
> Does calling Game::setLatCom(false) affect the other bot too? If that is the case (and it looks that way since it's a game setting and not a player setting) one cannot simply change this setting --> it would break the other bot.
> And Second ... that would mean that LatCom is applied to the enemy units too and my bot can "see" what enemy units are doing before the latency is over?

I see that you clarified this point, but I'll explain for anyone who doesn't know.
SC is a p2p game, and running Local PC does not create only one server. There are still two servers sending data to eachother. The BWAPI server is also separate to this and runs on the SC client. It's called a server only because this is what the API is talking to, not the game client.

> Furthermore I looked into your suggestion and it is of course doable but I don't like it because:
> It is more work and it's a disadvantage in some situations where the client won't accept commands due to latency. For example when cancelling a morph without LatCom you need to wait to get the resources refunded before you can build something else. With LatCom you can build immediately saving 1xLatencyFrames time. Yeah ... it's not much but more work for a disadvantage --> which is bad.

Yes but any bot could implement something that makes so that BWAPI (or other interface ) doesn't shoot down commands if you have insufficient resources, even players can do this.

> @N00byEdge You said you are going to fix this? :-)
> (I would of course volunteer as well but I don't know if it is even worth to fix with 5.0 coming up.)

Yes, I'm working on this. I don't see a reason latcom would be removed from BWAPI, seeing as it's part of the server, which might not change. If I end up getting latcom to a stable and easily patchable state, I see no reason to remove it from BWAPI. The change looks to be a ~1200 line commit of which I've written about 800 at this moment. I am hoping to get it finished late next week or the week after that.

**Comment by xatian** [comment]
> Yes but any bot could implement something that makes so that BWAPI (or other interface ) doesn't shoot down commands if you have insufficient resources, even players can do this.

I don't understand ... please clarify what you mean by that.
How can a player or a bot build something without sufficient resources?

> Yes, I'm working on this. I don't see a reason latcom would be removed from BWAPI, seeing as it's part of the server, which might not change. If I end up getting latcom to a stable and easily patchable state, I see no reason to remove it from BWAPI.

I don't think it should be removed either ... thank you for fixing it!

**Comment by N00byEdge** [comment]
>I don't understand ... please clarify what you mean by that.
>How can a player or a bot build something without sufficient resources?
Under the assumption that the resources will be available when the command is executed, that could be done. We can swap out latency for a worker building a building. You can queue up an SCV and make one build a supply depot when you only have 100 minerals, and everything will work just fine as long as you have mined 50 more minerals before the depot is placed.

**Comment by heinermann** [comment]
Should be fixed now after #783 but please reopen if that's not the case.
